New Serverless Pattern - REST API via API Gateway, acting as an S3 proxy (SLS & Python) (#2122)
aws-samples/serverless-patterns: apigw-s3-lambda-sls-py/README.md

# REST API as an Amazon S3 proxy in API Gateway

> This pattern utilizes Serverless framework to deploy a REST API via API Gateway, acting as an S3 proxy for write operations. 

Reviewer: ```suggestion
This pattern utilizes the Serverless framework to deploy a REST API via API Gateway, acting as an S3 proxy for write operations. 
```

> It includes API-Key Setup, Request Body Validator, Lambda invocation on S3 events, and enables CloudWatch logs and X-Ray tracing for API Gateway and Lambda using AWS Powertool for Lambda framework.

Reviewer: ```suggestion
It includes API-Key Setup, Request Body Validator, Lambda function invocation on S3 events, and enables CloudWatch logs and X-Ray tracing for API Gateway and Lambda using the Powertools for AWS Lambda.
```

Reviewer: Please include a link to Powertools

* [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) installed
* [Python 3](https://www.python.org/downloads/) installed
* [Serverless Framework](https://www.serverless.com/framework/docs/getting-started) installed
* [Powertools for AWS Lambda (Python)](https://docs.powertools.aws.dev/lambda/python/latest/)

## Setup and Deployment Instructions

1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
    ```commandline
    git clone https://github.com/aws-samples/serverless-patterns
    ```
2. Change directory to the pattern directory:
    ```commandline
    cd apigw-s3-lambda-sls-py
    ```
3. Install dependencies:
    ```commandline
    pipenv install
    ```
4. Install following Serverless (sls) Framework plugins:
    ```commandline
    sls plugin install -n serverless-python-requirements
    ```
   - `serverless-python-requirements` plugin will automatically detect `Pipfile` and pipenv installed and generate a `requirements.txt` file

5. Deploy the services specified in [serverless.yml](./serverless-compose.yml) by running:
   ```commandline
      sls deploy --stage dev --verbose
   ```

> 6. The outputs from the Serverless deployment process containing the resource names, IDs and/or ARNs which are used for

Reviewer: ```suggestion
6. The outputs from the Serverless deployment process contain the resource names, IDs and/or ARNs which are used for
```

   testing.

## How it works

The high-level diagram below serves to visually represent this pattern.

![apigw-s3-lambda-sls-py.png](./docs/apigw-s3-lambda-sls-py.png)

> This setup involves creating and launching a REST API using API Gateway, allowing direct S3 writing without additional services. 

Reviewer: ```suggestion
This setup involves creating and launching a REST API using API Gateway, allowing direct writing to an Amazon S3 bucket without additional services. 
```

> It utilizes an open-source Serverless framework for deployment. The pattern includes setting up API keys with Usage Plan, Request Body Validator, and triggering Lambda functions on S3 events.

Reviewer: ```suggestion
It utilizes the open-source Serverless framework for deployment. The pattern includes setting up API keys with Usage Plan, Request Body Validator, and triggering Lambda functions on S3 events.
```

> Additionally, this setup establishes Cloudwatch logs and X-Ray tracing for both API Gateway and Lambda, harnessing the capabilities of AWS Powertool specifically for Lambda functions.

Reviewer: ```suggestion
Additionally, this setup establishes Cloudwatch logs and X-Ray tracing for both API Gateway and Lambda, harnessing the capabilities of Powertools specifically for Lambda functions.
```

## Testing

#### I. Happy Path

This setup can be tested using the [curl command](https://github.com/curl/curl/blob/master/docs/MANUAL.md) or  [xh utility](https://github.com/ducaale/xh)  by accessing the `PUT /dev/api/{order_object_path+}
` endpoint. 

In this context, the inclusion of 'suffix' alongside 'order_object_path' suggests that we have the flexibility to substitute multiple folders in place of 'order_object_path'.

```commandline

> xh put https://${ApiGatewayRestApi}.execute-api.${AWS::Region}.amazonaws.com/${sls:stage}/api/orders_recevied/order001.json X-Api-Key:ttLMU7hv2S4qEpBVaCYXt4OWoy38jRmF7txgEBXy orderDate=2024-02-06 orderPaymentAmount:=23.99 orderPaymentCurrency=USD customerId:=10002

Reviewer: Replace the API Key with a placeholer

```
_Or_
```commandline

curl --request PUT \
  --url https://${ApiGatewayRestApi}.execute-api.${AWS::Region}.amazonaws.com/${sls:stage}/api/orders_recevied/order001.json \
  --header 'x-api-key: ttLMU7hv2S4qEpBVaCYXt4OWoy38jRmF7txgEBXy' \
  --data '{
		"orderId": 10023,
		"orderDate": "2024-02-03",
		"orderPaymentAmount": 23.99,
		"orderPaymentCurrency": "USD",
		"customerId": 100023
}'
```
Response:

```commandline
HTTP/2.0 200 OK
content-length: 0
content-type: application/json
date: Wed, 07 Feb 2024 01:54:37 GMT
x-amz-apigw-id: SvhitG7XoAMEifg=
x-amzn-requestid: 1dd4f2f1-f5ac-46d6-8f65-0d60d4370f0b
x-amzn-trace-id: Root=1-65c2e2dd-737319bc469770c954e78a1c

```

#### II. Testing Request Validator:
In this scenario, a value of "ZZZ" is passed for `orderPaymentCurrency` in the request body, which is not among the permitted values.

```commandline

> xh put https://${ApiGatewayRestApi}.execute-api.${AWS::Region}.amazonaws.com/${sls:stage}/api/orders_recevied/order001.json X-Api-Key:ttLMU7hv2S4qEpBVaCYXt4OWoy38jRmF7txgEBXy orderDate=2024-02-06 orderPaymentAmount:=23.99 orderPaymentCurrency=ZZZ customerId:=10002

Reviewer: Replace API Key with a placeholder
